test_generators/README.md

```diff
@@ -0,0 +1,170 @@
+
+```bash
+# From the root of the generator directory:
+# Create a virtual environment (any venv/.venv/.venvs is git-ignored)
+python3 -m venv venv
+# Activate the venv, this is where dependencies are installed for the generator
+. venv/bin/activate
+```
+
+Now that you have a virtual environment, write your generator.
+It's recommended to extend the base-generator.
+
+Create a `requirements.txt` in the root of your generator directory:
+```
+eth-utils==1.4.1
+../../test_libs/gen_helpers
+../../test_libs/config_helpers
+../../test_libs/pyspec
+```
+The config helper and pyspec is optional, but preferred. We encourage generators to derive tests from the spec itself, to prevent code duplication and outdated tests.
+Applying configurations to the spec is easy, and enables you to create test suites with different contexts.
+
+Note: make sure to run `make pyspec` from the root of the specs repository, to build the pyspec requirement.
+
+Install all the necessary requirements (re-run when you add more):
+```bash
+pip3 install -r requirements.txt
+```
+
+And write your initial test generator, extending the base generator:
+
+Write a `main.py` file, here's an example:
+
+```python
+from gen_base import gen_runner, gen_suite, gen_typing
+
+from eth_utils import (
+    to_dict, to_tuple
+)
+
+from preset_loader import loader
+from eth2spec.phase0 import spec
+
+@to_dict
+def example_test_case(v: int):
+    yield "spec_SHARD_COUNT", spec.SHARD_COUNT
+    yield "example", v
+
+
+@to_tuple
+def generate_example_test_cases():
+    for i in range(10):
+        yield example_test_case(i)
+
+
+def example_minimal_suite(configs_path: str) -> gen_typing.TestSuiteOutput:
+    presets = loader.load_presets(configs_path, 'minimal')
+    spec.apply_constants_preset(presets)
+
+    return ("mini", "core", gen_suite.render_suite(
+        title="example_minimal",
+        summary="Minimal example suite, testing bar.",
+        forks_timeline="testing",
+        forks=["phase0"],
+        config="minimal",
+        handler="main",
+        test_cases=generate_example_test_cases()))
+
+
+def example_mainnet_suite(configs_path: str) -> gen_typing.TestSuiteOutput:
+    presets = loader.load_presets(configs_path, 'mainnet')
+    spec.apply_constants_preset(presets)
+
+    return ("full", "core", gen_suite.render_suite(
+        title="example_main_net",
+        summary="Main net based example suite.",
+        forks_timeline= "mainnet",
+        forks=["phase0"],
+        config="testing",
+        handler="main",
+        test_cases=generate_example_test_cases()))
+
+
+if __name__ == "__main__":
+    gen_runner.run_generator("example", [example_minimal_suite, example_mainnet_suite])
+```
+
+Recommendations:
+- you can have more than just 1 suite creator, e.g. ` gen_runner.run_generator("foo", [bar_test_suite, abc_test_suite, example_test_suite])`
+- you can concatenate lists of test cases, if you don't want to split it up in suites.
```

Comment:
I'd notice this
```suggestion
- you can concatenate lists of test cases, if you don't want to split it up in suites, however make sure they could be run with one handler.
```

Reply:
̀make sure the base chain constants are the same`?